File: Version3.0/SimiCartPluginFW/SimiFramework/Model/SimiModelCollection.m
//
//  SimiModelCollection.m
//  SimiCartPluginFW
//
//  Created by Tan Hoang on 2/8/14.
//  Copyright (c) 2014 SimiTeam. All rights reserved.
//

#import "SimiModelCollection.h"
#import "SimiModel.h"

@implementation SimiModelCollection

+ (SimiAPI *)getSimiAPI
{
    NSString *klass = [NSStringFromClass(self) stringByReplacingOccurrencesOfString:@"ModelCollection" withString:@"API"];
    Class api = NSClassFromString(klass);
    Class loopClass = [self superclass];
    while (api == nil && loopClass) {
        api = NSClassFromString([[loopClass description] stringByReplacingOccurrencesOfString:@"ModelCollection" withString:@"API"]);
        loopClass = [loopClass superclass];
    }
    return [api new];
}

- (id)init{
    self = [super init];
    if (self) {
        currentNotificationName = @"DidFinishRequest";
    }
    return self;
}

- (instancetype)initWithArray:(NSArray *)array{
    NSMutableArray *temp = [[NSMutableArray alloc] init];
    for (NSDictionary *obj in array) {
        SimiModel *model = [self getModel];
        [model setData:obj];
        [temp addObject:model];
    }
    self = [super initWithArray:temp];
    return self;
}

- (void)setData:(SimiMutableArray *)data{
    if ([self count]) {
        [self removeAllObjects];
    }
    NSMutableArray *temp = [[NSMutableArray alloc] init];
    for (NSDictionary *obj in data) {
        SimiModel *model = [self getModel];
        [model setData:obj];
        [temp addObject:model];
    }
    [self addObjectsFromArray:temp];
}

- (void)addData:(SimiMutableArray *)data{
    NSMutableArray *temp = [[NSMutableArray alloc] init];
    for (NSDictionary *obj in data) {
        SimiModel *model = [self getModel];
        [model setData:obj];
        [temp addObject:model];
    }
    [self addObjectsFromArray:temp];
}

- (SimiModel *)getModel{
    Class superClass = self.class;
    Class modelClass = NSClassFromString([[superClass description] stringByReplacingOccurrencesOfString:@"Collection" withString:@""]);
    while (modelClass == nil) {
        superClass = [superClass superclass];
        modelClass = NSClassFromString([[superClass description] stringByReplacingOccurrencesOfString:@"Collection" withString:@""]);
    }
    return [modelClass new];
}

- (SimiAPI *)getAPI{
    return [self.class getSimiAPI];
}

- (NSString *)currentNotificationName
{
    return currentNotificationName;
}

- (void)didFinishRequest:(NSObject *)responseObject responder:(SimiResponder *)responder{
    if (responder.simiObjectName) {
        currentNotificationName = responder.simiObjectName;
    }
    [[NSNotificationCenter defaultCenter]postNotificationName:@"TimeLoaderStop" object:currentNotificationName];
    if ([responseObject isKindOfClass:[SimiMutableArray class]]) {
        switch (modelActionType) {
            case ModelActionTypeInsert:{
                [self addData:(SimiMutableArray *)responseObject];
            }
                break;
            default:{
                [self setData:(SimiMutableArray *)responseObject];
            }
                break;
        }
        [[NSNotificationCenter defaultCenter] postNotificationName:currentNotificationName object:self userInfo:@{@"responder":responder}];
    }else if (responseObject == nil){
        [[NSNotificationCenter defaultCenter] postNotificationName:currentNotificationName object:self userInfo:@{@"responder":responder}];
    }
}

- (void)preDoRequest{
    [[NSNotificationCenter defaultCenter] postNotificationName:[NSString stringWithFormat:@"%@Before", currentNotificationName] object:self];
    [[NSNotificationCenter defaultCenter]postNotificationName:@"TimeLoaderStart" object:currentNotificationName];
}

@end

File: simitracking/simitracking/SimiFramework/Model/SimiModelCollection.swift
//
//  SimiModelCollection.swift
//  SimiTracking
//
//  Created by Hai Nguyen on 11/14/16.
//  Copyright © 2016 SimiCart. All rights reserved.
//

import UIKit

class SimiModelCollection: NSObject {
    
    static var simiAPI:SimiAPI!
    public var currentNotificationName:String = "DidFinishRequest"
    public var data:Array<Dictionary<String, Any>> = []
    
    public var error:Array<Dictionary<String, Any>>!
    public var isSucess:Bool!
    
    public var allIds:Array<String>?
    public var total:Int?
    public var pageSize:Int?
    public var from:Int?
    public var responseObject:Dictionary<String, AnyObject>?
    
    
    public func getAPI()->SimiAPI{
        return SimiAPI()
    }
    
    public func deleteData() {
        data.removeAll()
    }
    
    public func getResource()->String {
        return ""
    }
    
    public func setData(data: Array<Dictionary<String, Any>>) {
        if (self.data.count>=1) {
            self.data.removeAll()
        }
        for item in data {
            self.data.append(item)
        }
    }
    
    public func addData(data: Array<Dictionary<String, Any>>) {
        for item in data {
            self.data.append(item)
        }
    }
    
    func preDoRequest() {
        NotificationCenter.default.post(name: Notification.Name(rawValue: currentNotificationName+"Before"), object: self)
        NotificationCenter.default.post(name: Notification.Name(rawValue: "TimeLoaderStart"), object: currentNotificationName)
        if (currentNotificationName != "DidFinishRequest") {
            NotificationCenter.default.post(name: Notification.Name(rawValue: "DidFinishRequest-Before"), object: self)
        }
    }
    
    public func didFinishRequest(responseObject:Dictionary<String, AnyObject>){
        self.error = nil
        if responseObject["errors"] != nil
        {
            self.error = responseObject["errors"] as! Array<Dictionary<String, Any>>
            self.isSucess = false
        }
        else {
            self.isSucess = true
            allIds = responseObject["all_ids"] as! Array<String>?
            total = responseObject["total"] as? Int
            pageSize = responseObject["page_size"] as? Int
            from = responseObject["from"] as? Int
            self.responseObject = responseObject
        }
        if self.isSucess == true {
            setData(data: responseObject[getResource()] as! Array<Dictionary<String, Any>>)
        }
        self .pushNotifications(responseObject: responseObject)
    }
    
    public func pushNotifications(responseObject:Dictionary<String, AnyObject>) {
        NotificationCenter.default.post(name: Notification.Name(rawValue: currentNotificationName+"After"), object: self)
        NotificationCenter.default.post(name: Notification.Name(rawValue: "TimeLoaderStop"), object: currentNotificationName)
        NotificationCenter.default.post(name: Notification.Name(rawValue: currentNotificationName), object: self, userInfo: responseObject)
        
        if (currentNotificationName != "DidFinishRequest") {
            NotificationCenter.default.post(name: Notification.Name(rawValue: "DidFinishRequest-After"), object: self)
        }
    }
}
